plugins/chat/app/queries/chat_channel_unreads_query.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class ChatChannelUnreadsQuery
+  def self.call(channel_id, user_id)
+    builder = DB.build(<<~SQL)
+      SELECT (
+      SELECT COUNT(*) AS unread_count
+      FROM chat_messages
+      INNER JOIN chat_channels ON chat_channels.id = chat_messages.chat_channel_id
+      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = chat_channels.id
+      WHERE chat_channels.id = :channel_id
+      AND chat_messages.user_id != :user_id
+      AND user_chat_channel_memberships.user_id = :user_id
+      AND chat_messages.id > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
+      AND chat_messages.deleted_at IS NULL
+    ) AS unread_count,
+    (
+      SELECT COUNT(*) AS mention_count
+      FROM notifications
+      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = :channel_id
+      AND user_chat_channel_memberships.user_id = :user_id
+      WHERE NOT read
+      AND notifications.user_id = :user_id
+      AND notifications.notification_type = :notification_type
+      AND (data::json->>'chat_message_id')::bigint > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
+      AND (data::json->>'chat_channel_id')::bigint = :channel_id
+    ) AS mention_count;
+    SQL
+
+    builder
```

Comment:
I'm not sure all this is necessary...couldn't you just do this?

```
sql =
  <<~SQL,
      SELECT (
      SELECT COUNT(*) AS unread_count
      FROM chat_messages
      INNER JOIN chat_channels ON chat_channels.id = chat_messages.chat_channel_id
      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = chat_channels.id
      WHERE chat_channels.id = :channel_id
      AND chat_messages.user_id != :user_id
      AND user_chat_channel_memberships.user_id = :user_id
      AND chat_messages.id > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
      AND chat_messages.deleted_at IS NULL
    ) AS unread_count,
    (
      SELECT COUNT(*) AS mention_count
      FROM notifications
      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = :channel_id
      AND user_chat_channel_memberships.user_id = :user_id
      WHERE NOT read
      AND notifications.user_id = :user_id
      AND notifications.notification_type = :notification_type
      AND (data::json->>'chat_message_id')::bigint > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
      AND (data::json->>'chat_channel_id')::bigint = :channel_id
    ) AS mention_count;
    SQL
  DB
    .query(
      sql,
      channel_id: channel_id,
      user_id: user_id,
      notification_type: Notification.types[:chat_mention],
    )
    .first
    .to_h
```

This will always return `{:unread_count=>0, :mention_count=>0}` (with the correct counts, and 0 if there really are none)

Reply:
There is no need for `query_hash` , safe nav, or `symbolize_keys` then

Reply:
Yes sure cool :+1: I tried something like this but didn't get what I wanted, must have done an error. THanks.